### PR TITLE
Cleaning Up UIItem

### PR DESCRIPTION
**Pull request by dpisanu**
Cleaning Up UIItem and IUIItem

- Adding Regions for sets of Functionality such as 'Interaction with an
UI Item', 'Properties of an UI Item', etc
- Renamed the RaiseClickEvent to Invoke
- Added code documentation
- Moved some public properties and functions also into the Interface to
get the Interface and implementing class in Synch

- Renamed UIItemEventListener to IUIItemEventListener

!! Migration Note !!
- Renamed the IUIItem.RaiseClickEvent to IUIItem.Invoke
- Renamed UIItemEventListener  to IUIItemEventListener

### CODE REVIEW

**Comment by Roemer** [comment]
Please add migration notes to the releasenodes file. I will review the code later.

File: src/TestStack.White/UIItems/IUIItem.cs [hunk 1]
```diff
@@ -86,32 +237,63 @@ public interface IUIItem : IActionListener
...
+        /// </summary>
+        /// <param name="property">The <see cref="AutomationProperty"/> value to compare to</param>
+        /// <param name="compareTo">The object to compare it to</param>
+        /// <returns>Returns <c>true</c> if matches, else <c>false</c></returns>
+        bool ValueOfEquals(AutomationProperty property, object compareTo);
 
-        void DrawHighlight();
-        void DrawHighlight(Color color);
+        /// <summary>
+        /// CHeck if the UI Item equals an object
```
Comment (Roemer) [inline src/TestStack.White/UIItems/IUIItem.cs]: Uppercase H in CHeck ;)

Comment (dpisanu) [inline src/TestStack.White/UIItems/IUIItem.cs]: Nitpicking :p

**Comment by dpisanu** [comment]
Done

File: src/TestStack.White/UIItems/IUIItem.cs [hunk 2]
```diff
@@ -11,71 +12,222 @@
...
+        void RightClick();
+
+        /// <summary>
+        /// Performas a right mouse click at a defined <remarks>Point</remarks>
+        /// </summary>
+        /// <param name="point">Point to mouse right click</param>
+        void RightClickAt(Point point);
+
+        /// <summary>
+        ///Perform keyboard action on this UI Item
```
Comment (Roemer) [inline src/TestStack.White/UIItems/IUIItem.cs]: Missing space ;)

File: src/TestStack.White/UIItems/UIItem.cs [hunk 3]
```diff
@@ -421,50 +540,97 @@ protected virtual void EnterData(string value)
...
-        /// <summary>
-        /// Highlight UIItem with a colored frame
-        /// </summary>
-        public virtual void DrawHighlight(Color color)
+
+        #endregion
+        
+        #region Private
+
+        private void RightClickOnCenter()
```
Comment (Roemer) [inline src/TestStack.White/UIItems/UIItem.cs]: Remove this method, it's only one line and called once.

**Comment by Roemer** [comment]
I like your refactoring style, feel free to add more PRs.

**Comment by JakeGinnivan** [comment]
Why make the breaking change `Renamed the RaiseClickEvent to Invoke`

And is adding an `I` to the interface worth potentially giving people compilation issues when they upgrade..

**Comment by dpisanu** [comment]
I feel RaiseClickEvent is the wrong naming when working with the Invoke Pattern.
It suggests an actual Click is being done, which is wrong.
I have had controls where the InvokePattern pypasses any Click Eventing / Handling.
So this is an attempt to call it what it is. Doing an Invoke on the UI Item and not a click.

I just stumbled upon it while doing code documentation. If I had known this existed earlier  would not have had to implement it manually again in my wrappers in white.
That just to show that names can be missleading.

Why to the Interface change, look at my answer in the different thread.

In a side note though. There will always be the case that further development will break integration.
That is why you document such changes and thenintegrating party does integration testing.

**Comment by JakeGinnivan** [comment]
Do you follow SemVer in your other projects?

What about when White is a transitive dependency, these sort of changes require re-releasing intermediate libraries and make the ecosystem around a open source library much more complex. 

Why not *add* the `Invoke` method, then mark `RaiseClickEvent` as obsolete. Then you improve the public API and give people time to transition.

**Comment by JakeGinnivan** [comment]
And the interface change, IMO is that White is an old project with quite a lot of people using it. It is not *that* important that we are following the guidelines, adding the I really doesn't improve anything or add value but it could mean that people who upgrade now have to spend time to fix the errors.

Personally I try *really* hard not to make breaking changes to any open source library I maintain, in Shouldly I have a test which fails when the public API changes so I must review the change and make sure it is a non-breaking change.

**Comment by dpisanu** [comment]
We do try to follow SemVer in our projects. 
Its always a battle though. As I am sure you know as well.

Not aware of Whites guidelines though.

I mentioned marking things as Deprecated in the other thread. I didnt get any feedback to it though.
So i went the way of documenting it in the Changelog.

I see the need of a guideline.

**Comment by JakeGinnivan** [comment]
> Its always a battle though. As I am sure you know as well.

Yes it is, in the case we have not even bothered though.

While we don't have contributing guidelines, or decent docs maybe err on the side of not breaking things? Lets provide some value first, get some releases which fix things/improve things for *consumers* of the library, then start shooting for v1 where we can remove any deprecated APIs and make all breaking changes in a single release which declares the White public API. Then we can follow SemVer from there.

**Comment by Roemer** [comment]
Why not shooting for v1 right now? I mean there hasn't been a release for ages and it seems the current release works fairly ok für most. I think we should try to get the UIACom branch working.

**Pull request merged**